### DISCUSSION TITLE
Does Microdot support http version 1.1 ?

### DISCUSSION THREAD

**Discussion by lukaskremla** [miguelgrinberg/microdot#246] [General]
As the title suggests. I have noticed that my files are served with http 1.0 through Microdot send_file method . The web application I am working with is quite large for a microcontroller and using http 1.1 could speed up the load times. I haven't been able to find much about http version configuration from the documentation besides this:

class microdot.Request(app, client_addr, method, url, http_version, headers, body=None, stream=None, sock=None)

But the documentation doesn't mention anything about the http_version parameter and I couldn't find any usages of it in the source code to learn from either.

Does Microdot support http 1.1?

**Comment by miguelgrinberg** [comment]
Microdot is a web framework, not a web server, so it does not care what HTTP version is used in HTTP exchanges. The web server that is included in this repository is minimalistic by design. If there is a different web server that you would like to use, then it should be possible to build an integration between it and Microdot. On CPython I have built integrations with the WSGI and ASGI web server standards.

**Comment by beyonlo** [comment]
@Luky557 As suggestion, if you need a large web application with Microdot (where you can't to use another Web Server as @miguelgrinberg suggest you - example:  using Microdot in microcontrollers), a good alternative is write all your web application in JavaScript using Websocket communication between your JS frontend and the Microdot WebSocket Server. So you have just one load time to get all JS from Server to Client (browser), and after that you just need to send and receive JSON messages between JS application e Microdot over WebSocket.

**Comment by lukaskremla** [comment]
Thank you both for the replies. 

I have looked for alternatives as far as whole MicroPython packages (MicroWebSrv2 for example) go and Microdot is still the best option. MicroWebSrv2 works with http 1.1, but the documentation is fairly poorly written (at least in comparison to Microdot) and it lacks functionality like Microdot's "send_file" and it has no out of the box gzip support. I could modify their code to work for what I need, but unless I would dive into a full rewrite, these would be poor workarounds and I would eventually end up with a hard to manage and maintain codebase. 

The web socket and JS solution is sort of what I am doing already. The web application I am running is quite large and it has its own lazy-loading. The way it works is that as soon as you land on the main page everything gets downloaded in the background. Thing is, each time this web app source code is generated it is split into many chunks and their names vary, I can't hard code the web sockets to send all of these files to the browser to each new session. The first "/" request serves the index which then ends up with everything else being served after in my web server code as the requests come in. I don't think I could benefit from web sockets here in anyway, but please do correct me if I miss understood something, I am by no means a well experienced dev, happy to learn new things.

I am working on an ESP32 S2 with a flash of 4 MBs. Could I please ask for some directions or tips as far as modifying the web server included with Microdot to support http 1.1? Is that something that is even viable? Any chance I would find a different web server capable of supporting the current Microdot functionality, http 1.1 which would fit on the 4 MB flash that I could try to integrate with Microdot?

You are much more experienced than me, I would welcome any directions or tips for study material on this. Thank you for the replies so far!

**Reply by miguelgrinberg** [reply]
What feature from HTTP 1.1 do you need specifically?

**Reply by lukaskremla** [reply]
Persistent connections and pipelining are what I am after. Any loading speed increase I could achieve on the server is very welcome.

**Reply by miguelgrinberg** [reply]
My understanding is that pipelining isn't supported in modern browsers, there are newer mechanisms such as request multiplexing introduced in HTTP/2.

Persistent connections should be possible to implement in MicroPython is this does not require a lot of extra logic. The feature is described in the [HTTP specification](https://datatracker.ietf.org/doc/html/rfc7230#section-6.3). The gist of it is that the server should keep sockets around after they were used in a request, because the client might reuse them to send another request.

**Reply by lukaskremla** [reply]
Thanks for the directions and tips! 

The fact that pipelining is no longer supported saves me time at least. I will do some reading up on persistent connections and then I'll have a shot at getting it working in code. 

Thanks again for your help!
